frontend/src/Components/ScaleSelector/ScaleSelector.tsx

```diff
@@ -1,25 +1,128 @@
+import { useEffect, useRef, useState } from 'preact/hooks';
 import { ScaleName, SCALES } from '../../../../shared/scales';
 import { SELECT_CHANGE_SCALE } from '../../constants';
+import sharedClasses from '../../styles.module.css';
 import { connectToWebSocket } from '../WebSocket/WebSocket';
 import classes from './ScaleSelector.module.css';
 
-export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => (
-  <select
-    disabled={!connected}
-    name="scale"
-    class={classes.select}
-    onChange={({ target }) =>
-      setScale(SCALES[(target as HTMLSelectElement).value as ScaleName].values)
+const availableScales = Object.entries(SCALES);
+
+export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => {
+  const [open, setOpen] = useState(false);
+  const [selected, setSelected] = useState(-1);
+  const selectionButtonRef = useRef<HTMLButtonElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+
+  const close = () => {
+    setOpen(false);
+    setSelected(-1);
+  };
+
+  const toggle = () => {
+    if (open) {
+      close();
+    } else {
+      setSelected(0);
+      setOpen(true);
+    }
+  };
+
+  const handleBlur = ({ relatedTarget }: FocusEvent) => {
+    if (relatedTarget !== selectionButtonRef.current && relatedTarget !== dropdownRef.current) {
+      close();
+    }
+  };
+
+  const handleKeyboardNavigation = (event: KeyboardEvent) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        setSelected((selected + 1) % availableScales.length);
```

Comment:
I would generally prefer 

```suggestion
        setSelected((selected) => (selected + 1) % availableScales.length);
```
and a similar change for `ArrowUp` since these keys can be pressed rapidly and we don't want to suppress events while re-rendering. However, I myself didn't notice any impairment when testing the behavior on my machine.

Reply:
Ah, I actually did not know this, good idea!